SILBuilder ergonomics for improved debug scope handling

vedantk opened the pull request:
SILBuilder is designed to be flexible and simple, but its current design make it a bit too easy to emit instructions with the wrong debug scopes. The goal of this PR is to make SILBuilder even easier to work with, but harder to misuse. Hopefully this will set us up for improved debug info handling in the compiler.

A common reason why the wrong debug scope is applied to an instruction is because a SILBuilder's insertion point (IP) was changed without a corresponding change to the DebugScope. There are a few ways to address this:

- _SILBuilderForCodeExpansion_: Introduce a dedicated SILBuilder variant which always sets the correct DebugScope when setting/changing its IP. The intended use is for expanding an existing instruction sequence: there are constructors available which automatically pick the right DebugScope for an insertion point.

- _Restrict setInsertionPoint()_: Restricting when setInsertionPoint can be used would make it harder to write a class of bugs. We should explore deleting the inherited setInsertionPoint methods in SILBuilderForCodeExpansion, and/or adding strict asserts to the originals in SILBuilder.

- _Discourage SILBuilder recycling_: We might discourage reuse of a single SILBuilder instance within passes which rewrite or update SIL. It isn't necessary to go to great lengths to avoid constructing a SILBuilder, e.g by repeatedly changing a single builder's IP. It is cheap (and probably better from a correctness standpoint) to simply construct fresh builders as needed. There's more discussion needed here as I'm not sure whether SILBuilder reuse is more idiomatic. I'm also not sure how to discourage SILBuilder recycling, other than by introducing stricter asserts in setInsertionPoint and by pointing the issue out in code review.

- _A SILBuilder variant for code movement_: We might consider adding a third SILBuilder variant intended to be used for code movement only. The idea would be to inherit a debug scope from the original position of the code. For now, I've left this out of the PR, because of the overlap in use cases with the existing builder variants.

I'm open to adjusting these approaches or considering new ones.

rdar://35575482

vedantk committed:
Rename SILBuilder{WithScope -> ForCodeExpansion}, NFC

SILBuilderForCodeExpansion should be used to expand a sequence of
instructions. The new name better communicates how this version of
SILBuilder should be used.

On construction, SILBuilderForCodeExpansion infers the current debug
location and scope from the specified insertion point. The goal is to
make it harder to accidentally set the wrong debug scope when generating
instructions.

There's more we can do to work towards this goal --

I have follow-up commits planned which will add asserts to both
SILBuilder and SILBuilderForCodeExpansion, to make sure that they are
each used for their intended purposes.

For example, it should not be possible to change the insertion point of
a SILBuilderForCodeExpansion, because this might invalidate the debug
scope. As another example, setting the insertion point of a SILBuilder
in between two instructions should not result in non-linear stepping
behavior in the debugger.

Part of: r://35575482

vedantk committed:
Simplify a SILBuilderForCodeExpansion constructor

This simplifies the SILBuilderForCodeExpansion constructor which
positions the builder at the start of a basic block.

First, I made the constructor static, to make it easier to tell it apart
from the position-at-end constructor.

Next, I got rid of getScopeOfFirstNonMetaInstruction(). This function
did the opposite of what it claimed to do, but did not cause problems in
practice because it managed to preserve debug scopes on the instructions
its only client, the LoadableByAddress pass, visited.

It wasn't clear why we'd want to use the scope of the first non-meta
instruction when inserting at the start of a block, as this is a bit
heuristic, and not guaranteed to pick the right scope. The only solution
I found was to make the InheritScopeFrom argument mandatory here.

There are some oddities in this area. In the test case, you can see that
the debug scope assigned to the last three alloc_stack instructions
begins at the end of the function. This is something we need to examine
later. The same scope is assigned to these instructions before
LoadableByAddress runs, so I won't attempt to address the issue in this
patch in order to keep the diff simple.

As a follow-up, I plan on simplifying SILBuilderForCodeExpansion further
by deleting its inherited setInsertionPoint methods, which should
improve correctness w.r.t setting the right debug scope.

Part of: r://35575482

adrian-prantl commented on include/swift/SIL/SILBuilder.h:
```diff
@@ -2058,53 +2058,56 @@ class SILBuilder {
-/// SIL instructions.
-class SILBuilderWithScope : public SILBuilder {
+/// A wrapper on top of SILBuilder's constructor that sets the insertion point,
+/// and automatically sets the SILDebugScope based on the specified IP. This is
```
what's IP?

adrian-prantl commented on include/swift/SIL/SILBuilder.h:
```diff
@@ -2058,53 +2058,56 @@ class SILBuilder {
-class SILBuilderWithScope : public SILBuilder {
+/// A wrapper on top of SILBuilder's constructor that sets the insertion point,
+/// and automatically sets the SILDebugScope based on the specified IP. This is
+/// useful for situations where a single SIL instruction is lowered into a
```
I would be even more affirmative:

`Use this variant of SILBuilder when a single ...`

adrian-prantl commented on include/swift/SIL/SILBuilder.h:
```diff
@@ -2058,53 +2058,56 @@ class SILBuilder {
 
 public:
-  explicit SILBuilderWithScope(
+  explicit SILBuilderForCodeExpansion(
```
Would be nice to add a doxygen comment for the constructor that explains that InsertedInstrs returns a list of the instruction built by this builder etc.

adrian-prantl reviewed (commented):
Looks generally good, couple of inline comments.

vedantk committed:
Delete unsafe setIP methods within SILBuilderForCodeExpansion

Any method in SILBuilderForCodeExpansion which may inadvertently allow
the debug scope to become invalid is unsafe.

This deletes the unsafe setInsertionPoint() methods inherited by
SILBuilderForCodeExpansion. They are replaced with new overloads which
require the user to specify an instruction to inherit a debug scope
from, or which automatically infer the correct debug scope in the exact
same way the SILBuilderForCodeExpansion constructors do.

Part of: r://35575482

vedantk commented on include/swift/SIL/SILBuilder.h:
```diff
@@ -2058,54 +2058,121 @@ class SILBuilder {
+  }
+  static SILBuilderForCodeExpansion atBeginning(SILBasicBlock *BB) {
+    assert(!BB->empty() && "No code to expand in basic block");
+    return atBeginning(BB, &*BB->begin());
```
I think it would still be possible for getScopeOfFirstNonMetaInstruction() to pick up the wrong scope to apply. It might be better to require clients to be more explicit about which scope they want here.

vedantk commented:
SILBuilderForCodeExpansion::atBeginning now requires users to explicitly pass in an instruction to inherit a debug scope from, which should address the last piece of review feedback (so far).

vedantk commented:
@swift-ci Please test

gottesmm commented:
I am starting to read this code. A few thoughts:

1. I added Andy. He has opinions here I think.
2. *NOTE* I am not thinking about how this fits with current usage in the code base, but in the abstract: If we do not want people passing around SILBuilder and always constructing a new one then why not:
   a. Eliminate the copy operator/constructor. This just makes sense.
   b. Change all methods on SILBuilder to only be able to be invoked on r-value references and delete the move operator/constructor. This would mean that one would have to perform a std::move(...) on the value making it more difficult to use when it is stored in a structure and making its ephemeral nature much clearer.

Just some random thoughts.

atrick commented:
@gottesmm I haven't read this all yet. I just want to say that we definitely *do* want to reuse SILBuilder instances. That's how we track the current insertion point. The only reason we would want to change the insertion point, instead of creating a new SILBuilder, is if we have a reference `SILBuilder &`. If we always store/pass them by value, then there's no sense in changing the insertion point.

vedantk committed:
Disable copy construction/assignment for SILBuilder, NFC

As Michael Gottesman points out, there aren't any obvious reasons to
permit copying SILBuilder, as the preferred approach is to create fresh
builder instances as needed.

vedantk commented:
> Eliminate the copy operator/constructor. This just makes sense.

@gottesmm Done.

> Change all methods on SILBuilder to only be able to be invoked on r-value references and delete the move operator/constructor.

I like this idea. I wonder thought if it wouldn't make the builder harder to work with? Wouldn't this make it impossible to, say, write a helper which constructs & returns a builder?

We should also consider that this would be a much more involved refactoring. There are around two hundred sites which accept a SILBuilder as a reference, and I don't think I have the time to audit each / rewrite them to pass builders by value.

> I haven't read this all yet.

@atrick, sorry about the large diff. This PR should be easier to read using github's commit viewer (the first commit is the big NFC one, and is probably fine to skip).

> The only reason we would want to change the insertion point, instead of creating a new SILBuilder, is if we have a reference SILBuilder &. If we always store/pass them by value, then there's no sense in changing the insertion point.

As I mentioned above, I don't think I'll be able to make SILBuilder exclusively pass-by-value in this PR. It does seem like deleting setInsertionPoint within SILBuilderForCodeExpansion is a meaningful step in the right direction, so I'm hoping we can start with that.

vedantk commented:
Friendly ping. I'd like to know whether this is going in the right direction before rebasing.

atrick commented:
Why isn't a normal "SILBuilder" always the one that handles "code
expansion"? That's the one that will be created all over the
compiler. SILGen has one instance of a special builder that is
wrapped inside SILGenBuilder. It can have the most awkward name we can
imagine. The biggest problem with "SILBuilderWithScope" today is that
it's the one with the special name. I have never understood why a
normal SILBuilder didn't enforce a DebugScope.

I think you need to decide whether SILBuilder is meant to be used
polymorphically or as a safely constructed value. Presumably, all
calls to B.setInsertionPoint(I) can be replaced with B = SILBuilder(I)
by making SILBuilder final. There may be places we still want to use
SILBuilder polymorphically. Those can take a SILBuilderBase, which
cannot be instantiated and cannot have its insertion point
modified--it just builds the instructions you ask in the scope and
insertion point provided the caller.

I agree with the stated goals of this PR, but don't see how renaming
SILBuilderWithScope to SILBuilderForCodeExpansion is a step
forward. The SILBuilderWithScope name makes more sense to me,
although, as I said, it should just be SILBuilder.

I don't understand why code motion vs. expansion has anything to do
with the debug scope, so I could be fundamentally missing
something. My understanding is that instructions should take on the
surrounding debug scope regardless of whether they are being moved or
inserted. Is that incorrect? I think that is the main issue that needs
to be clarified, either via the API or comments.

adrian-prantl commented:
> Why isn't a normal "SILBuilder" always the one that handles "code expansion"? 

Thanks for pointing this out, that's a really good point. There is no reason why the most common variant of SILBuilder needs to have the longest name.

> I don't understand why code motion vs. expansion has anything to do
> with the debug scope, so I could be fundamentally missing
> something. My understanding is that instructions should take on the
> surrounding debug scope regardless of whether they are being moved or
> inserted. Is that incorrect?

That is (usually) not correct. During most transformations we want instructions to keep their location and scope. When an instruction causes a trap then we want the crash log to point to the source location that caused the instruction to be generated. You are right in so far as there is tension between wanting accurate locations (for better crash logs and profiling) and smoother stepping in the debugger (so "next" doesn't jump back and forth in the program). In optimized code we usually tend to optimize for more accuracy, whereas in unoptimized we tend to optimize for smoothness. Sometimes, when an instruction doesn't correspond to anything the developer wrote, we let that instruction inherit the location of whatever is next to it to yield a smoother line table. This is probably the case that you had in mind.

I don't think that there is a case where an instruction's scope should ever be changed without also changing its location. For all practical purposes scope and location form an inseparable unit. (The one exception is inlining, because the scope happens to be the data structure where the inline history of an instruction is stored, but that's almost an implementation detail).

futurejones commented:
Can one of the admins verify this patch?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
